lib/noted/notes.ex

```diff
@@ -344,6 +359,63 @@ defmodule Noted.Notes do
     Tag.changeset(tag, attrs)
   end
 
+  def add_tag(user_id, note_id, tag_name) do
+
+    {:ok, tag} = get_tag_by_name(tag_name, user_id)
+
+    if (is_nil(tag) or empty_notes_tags?(note_id, tag.id)) do
+    {:ok, tags} = Repo.transaction(fn ->
+    result =
+        case tag do
+          nil ->
+            %Tag{}
+            |> Tag.changeset(%{name: tag_name, user_id: user_id})
+            |> Repo.insert()
+
+          tag ->
+            {:ok, tag}
+        end
+
+      case result do
+        {:ok, tag} ->
+          %NotesTags{}
+          |> NotesTags.changeset(%{note_id: note_id, tag_id: tag.id})
+          |> Repo.insert()
+
+        error ->
+          error
```

Comment:
This will not rollback the transaction. Try: `{:error, reason} -> Repo.rollback(reason)`

Reply:
rollback the txn changes https://github.com/lawik/noted/pull/2/commits/193e20f09eb22526394048e1a1697c159d923d5f